c++/src/kj/async-io.h

```diff
@@ -419,6 +419,10 @@ class ConnectionReceiver {
   // For backwards-compatibility, the default implementation of this method calls `accept()` and
   // then adds `UnknownPeerIdentity`.
 
+  virtual Promise<void> onFatalError();
```

Comment:
I like the `listen()` interface. I might suggest `kj::Promise<void> listen(ConnectionReceiver::Handler&)` instead, it gives us slightly more design space because we can expand the interface for `Handler` at will.

Reply:
Hmm, I somewhat agree, and note that it would save an allocation as well (kj::Function requires an allocation, but `Handler&` could point at the parent object).

However, the `kj::Function` approach is significantly less verbose for callers, since they don't have to define a class.

`kj::TaskSet`, for example, requires an `ErrorHandler&`, and my experience has been that it's actually pretty painful every time I want to create a `kj::TaskSet` and have to figure out what's going to satisfy the `ErrorHandler&`... in retrospect I wish I made it accept a `kj::Function`.

So I'm a bit torn here but I think I lean more towards the `kj::Function` approach.

Reply:
I guess maybe the `kj::Function`-based `listen()` could be provided as a non-virtual method that is implemented in terms of the `Handler&`-based `listen()`...

Reply:
Unfortunately, when I gave it a shot it really started looking like the proper behavior for mixed accept() and listen() would be painful to guarantee. On the upside, I think I've figured out a way to phrase your concerns on accept() exceptions into something that avoids a separate promise.

Reply:
FWIW I wouldn't expect to support mixed `accept()` and `listen()`. It's already the case that you can't have multiple overlapping calls to `accept()`, so also saying you can't have overlapping `listen()` seems reasonable.